xgcm/autogenerate.py

```diff
@@ -0,0 +1,254 @@
+
+        return ds
+
+
+def autogenerate_ds(ds,
+                    axes={'X': 'lon', 'Y': 'lat'},
+                    position='center',
+                    coord_axes=None,
+                    x_wrap=360,
+                    y_wrap=180,
+                    z_pad=None,
+                    x_coord_wrap=360,
+                    y_coord_wrap=180,
+                    z_coord_pad=None):
+    """
+    Regenerate all c-grid information from an existing dataset
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        The data on which to operate
+    axes : axes on which to regenerate dims (and optional coords, see
+        'coord_axes' input). Needs to be in the form of a dict with keys
+        ('X','Y'or'Z'), the values need to be the corresponding dimension names
+        in ds.
+    position : {'center', 'left', 'right'}
+        The position of the datapoints relative to the coordinates given
+        This will usually be the cell center, but can sometimes be the
+        lower left (option:'left') or perhaps upper right (option:'right')
+        corner of the gridcell.
+    coord_axes : (optional) Like 'axes' but for multidimensional coordinates in
+        ds
+    x_wrap : float - Value of discontinuity at the 'X' boundary
+    y_wrap : float - Value of discontinuity at the 'Y' boundary
+    z_pad : float -  Depth padding value. If None, it will be determined from
+        data
+    x_coord_wrap : float - Like 'x_wrap' but for coordinate variable
+    y_coord_wrap : float - Like 'y_wrap' but for coordinate variable
+    z_coord_pad : float -  Like 'z_pad' but for coordinate variable
+
+    Returns
+    -------
+    ds : xarray.DataArray
+        Dataset with inferred dims (and coordinates)
+    """
+
+    # set position of original coordinates and define how to interpolated
+    # For now I am assuming that observational datasets will use the left/right
+    # convention, rather then inner/outer...but that can be added.
+
+    # 'to' input maps the correct movement of coordinates if the input
+    # coord is set to center (this will be relabled appropriately below)
+    if position == 'center':
+        to = 'left'
+        to_shift = -0.5
+        from_shift = 0
+    elif position == 'right':
+        to = 'left'
+        to_shift = 0
+        from_shift = 0.5
+    elif position == 'left':
+        to = 'right'
+        to_shift = 0
+        from_shift = -0.5
+
+    # Do we want to modify in place? or copy?
+    ds = ds.copy()
+
+    # Loop over all specified axes and build new coordinates
+    ds = generate(ds, axes, axes, to, x_wrap, y_wrap,
+                  z_pad, to_shift, from_shift, dim_switch=True)
+
+    # Generate shifted multidimensional coordinates if specified in
+    # 'coord_axes'
+    if coord_axes:
+        ds = generate(ds, coord_axes, axes, to, x_coord_wrap, y_coord_wrap,
+                      z_coord_pad, to_shift, from_shift, dim_switch=False)
```

Comment:
It seems weird to be calling `generate` again. Can you explain what is going on here?